src/backend/distributed/executor/executor.c

```diff
@@ -995,21 +1022,177 @@ FindOrCreateWorkerSession(WorkerPool *workerPool, MultiConnection *connection)
 
 
 /*
- * RunDistributedExecution runs a distributed execution to completion. It
- * creates a wait event set to listen for events on any of the connections
- * and runs the connection state machine when a connection has an event.
+ * ShouldRunTasksSequentially returns true if each of the individual tasks
+ * should be executed one by one. Note that this is different than
+ * MultiShardConnectionType == SEQUENTIAL_CONNECTION case. In that case,
+ * running the tasks across the nodes in parallel is acceptable and implemented
+ * in that way.
+ *
+ * However, the executions that are qualified here would perform poorly if the
+ * tasks across the workers are executed in parallel. We currently qualify only
+ * one class of distributed queries here, multi-row INSERTs. If we do not enforce
+ * true sequential execution, concurrent multi-row upserts could easily form
+ * a distributed deadlock when the upserts touch the same rows.
+ */
+static bool
+ShouldRunTasksSequentially(DistributedPlan *plan)
+{
+	Job *workerJob = plan->workerJob;
+	List *taskList = workerJob->taskList;
+	Task *initialTask = NULL;
+
+	if (list_length(taskList) < 2)
+	{
+		/* single task plans are already qualified as sequential by definition */
+		return false;
+	}
+
+	/* all the tasks are the same, so we only look one */
+	initialTask = (Task *) linitial(taskList);
+	if (initialTask->rowValuesLists != NIL)
+	{
+		/*
+		 * We explicitly assert the fields to be NULL because we implemented the
+		 * single task plan executions for a restricted set of operations, e.g.,
+		 * only multi-row inserts. So, we don't expect other plans go through this
+		 * code, unless execution is implemented properly.
+		 */
+		Assert (plan->operation == CMD_INSERT);
+		Assert (plan->subPlanList == NIL);
+		Assert (plan->insertSelectSubquery == NULL);
+		Assert (plan->insertTargetList == NIL);
+		Assert (plan->intermediateResultIdPrefix == NULL);
+		Assert (plan->masterQuery == NULL);
+		Assert (workerJob->dependedJobList == NIL);
+
+		/* found a multi-row INSERT */
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * SequentialRunDistributedExecution gets a distributed execution and
```

Comment:
Just to share my thoughts for the reviewer and future reference: 
I think this function (and the logic in general) is pretty rigid. But, with the asserts and comments, I'm OK for now. If we want to make it more flexible in the future , we'd hit the asserts immediately. I don't want to over engineer this for cases that we currently don't support and will probably never need to support.